### ISSUE TITLE
Feature - Support AWS IAM Roles for Service Accounts

### ISSUE THREAD

**Issue by rifelpet** [kubernetes/kops#8264]
**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

AWS added support for IAM Roles using K8s service accounts, and I think it would be great if Kops could handle configuring this automatically. See the [blog post](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) and [eks docs](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html)

Kops can essentially follow the instructions here:

https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/SELF_HOSTED_SETUP.md

I have a WIP branch that will expose this through a set of API fields, but there are a few design decisions we'll need to make:

1. The public signing key needs to be publicly accessible through a URL. Options:
   - Upload it to the S3 state store bucket with an ACL allowing public read access. Having specific files be publicly accessible in a bucket that also contains sensitive objects could be concerning for users, or even disallowed by security requirements.
   - Have kops create a second bucket that is publicly accessible only used for hosting this file. This will require adding a new S3 bucket task to cloudup.
   - Host it elsewhere - API Gateway? (more new cloudup tasks) Or provide the file to the user for them to host however they want? The user would need to provide the resulting URL in the cluster spec so that Kops can add it to kube-apiserver's `--service-account-issuer`.

2. I think it would be good for Kops to create the OIDC provider for the cluster, since according to [this comment](https://github.com/aws/amazon-eks-pod-identity-webhook/issues/23#issuecomment-565109745) an OIDC provider can only be used with one cluster, so in most cases it would need to be created regardless. If a user wants to provide their own OIDC provider we could support that as well and optionally have Kops not create one. In my case, we'll be using an OIDC provider in a different AWS account as the cluster, but I dont think Kops needs to be concerned about that at all. 

3. It's not immediately clear to me, but I think the keypair that needs to be generated needs signing permissions, so should we reuse the existing cluster CA or create a new one just for this purpose?

4. This is minor, but I'm not sure exactly what we should name the API fields because I'm not entirely sure what the name of the feature is :) I see mentions of `IAM Roles for Service Accounts`, `Pod Identity Webhook`, I think the former is the more official name but might be tricky to fit in an API field name.

5. Theres already an [authentication.aws](https://godoc.org/k8s.io/kops/pkg/apis/kops#AwsAuthenticationSpec) API field in the cluster spec that provisions the [IAM Authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator) daemonset (used for IAM credential auth for k8s api clients), so I'm not sure if we should share that field space or create a new field. If we share the `authentication.aws` space we'd need to add an additional boolean to disable use of the IAM Authenticator, since it is enabled purely with `aws: {}` so we'd need a way to enable IAM Roles for Service Accounts while not enabling the AWS IAM Authenticator as well. This may also depend on if other cloud providers support or intend to support this functionality... do we make the field cloud-agnostic or have separate fields per provider?

**Comment by radarnex** [comment]
We're also interested in this feature

**Comment by johngmyers** [comment]
Could we not host the two documents from a Deployment on the cluster? Perhaps give an option to annotate the `Service` if we don't want to assume the cluster will support `Ingress`.

In general I think a separate keypair for a separate purpose would be preferable.

I would suggest a separate field than the one for the IAM Authenticator. It doesn't seem inherently AWS-specific feature. It looks like it might even be useful to things other than cloud providers.  Perhaps something like `serviceOIDCProvider`?

**Comment by rifelpet** [comment]
My concern with using a Deployment is for users who don't allow public ingress from the internet to their cluster. I don't think any solution could work for entirely air-gapped clusters, but I would hope we could at least find a solution for clusters that aren't allowed any public ingress.

I like the idea of a separate cloud-agnostic field in the API for this. We should keep it flexible so that down the road we allow more aspects of this configuration to be customizable. 

I'll keep thinking about potential solutions for hosting the public key...

**Comment by johngmyers** [comment]
I think the choices are one or more of:

* Deployment
* Second S3 bucket (plus equivalent option for other cloud providers)
* Provide file to be hosted externally

We might have to provide multiple options. There probably isn't a one-size-fits-all choice.

**Comment by rifelpet** [comment]
I've been giving this more thought. I think the MVP could provide the files within the state store and have the cluster admin host them publicly out of band. This allows us to later provide a hosting solution (manage an s3 bucket + objects, deployment, etc) in future iterations without locking us into any of those early. 

As an aside, there's an [implementable KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190730-oidc-discovery.md) to have kube-apiserver provide this information, so if it is implemented soon we may not need to deal with actually hosted these files and instead just providing access to them. I think a cluster with an internet-facing API ELB, public route53 hosted zone, and ACM cert would work automatically. Otherwise we'd need some other way of reaching the apiserver pods over the internet. Rather than having kops provision a special load balancer for this I wonder if it'd be easier to just let the cluster admin use their own ingress solution, ensuring it has a trusted cert and DNS and then have them provide the hostpath in the ClusterSpec. 

We need to decide whether kops should create an IDP or have the cluster-admin manage it, and if we want to support one of those options initially and add support for the other later. I think because the IDP is unique to the cluster (it can only have one issuer hostpath and that path will contain keys specific to the cluster) I don't see a benefit to having kops not manage the IDP. Perhaps an exception would be orgs not wanting to allow `iam:CreateOpenIDConnectProvider` permissions, but we can deal with that in later iterations. The IDP itself isn't relevant to anything else kops is concerned with, so we wouldn't need to provide the ARN in the ClusterSpec. Perhaps a boolean flag indicating whether or not kops should create the IDP would be sufficient.

That leaves the API design. We'll need to keep in mind the future functionality mentioned above and staying cloud-agnostic. I think it helps to list out every entity and their properties that will need to be managed either by kops or externally:

* Signing private and public keys - Issued by kops automatically
* Issuer HostPath - API field
* Audience - kops cluster name or API field
* discovery.json & keys.json
  * depends on: issuer hostpath, signing public key
* kube-apiserver flags
  * depends on: issuer hostpath, signing keys, audience
* AWS IDP (oidc)
  * depends on: issuer hostpath, audience 

AWS IDP's only restriction on audience value is <255 characters, so we could consider using the kops cluster name for this. Eventually we could allow this to be customized, perhaps when we also allow the IDP to be managed outside of kops.

It really seems like the issuer hostpath is the only field we need to accept for initial support, so perhaps something like this:

```yaml
type: Cluster
spec:
  serviceOIDCProvider:
    issuerHostPath: string
    # example values:
    # s3-us-east-1.amazonaws.com/my-oidc-bucket # if the cluster admin hosts the files in s3
    # ingress.cluster.example.com/my-oidc-path  # if using externally-managed ingress / dns / cert
    
    # future fields
    idp:
      create: false
    audience: foo
```

If we wanted to eventually support kops creating a public s3 bucket to host the json files, that would make the issuer hostpath a "computed" field so we would need to do validation to make sure it isn't specified if the s3 bucket is being created as well.

Thoughts on this api design and field naming?

**Comment by johngmyers** [comment]
I would prefer an `issuer` field that was an URL over an `issuerHostPath`.

How are you planning on getting the thumbprint(s) needed for the `CreateOpenIDConnectProvider` call?

Having `idp.create` be down a level seems wonky.

**Comment by rifelpet** [comment]
Yes `issuer` makes more sense, every use of includes the protocol so it would be more intuitive to accept a URL.

Heres my work so far: https://github.com/kubernetes/kops/compare/master...rifelpet:irsa?expand=1

My current roadblock is indeed getting the thumbprint from a KeyPair task to the OIDCProvider task. I don't see any other cases of tasks depending on the output of other tasks beyond just retrying cloud provider api calls for eventual consistency. I tried passing the CAStore to the task but I dont know if that will be problematic, and at least the [integration tests fail](https://travis-ci.org/github/rifelpet/kops/builds/680124392) because it can never find the key in the keystore. I haven't tried it live against AWS yet though.

Maybe theres an easier way to access the CAStore from a task?

I think the remaining items are:
* Figure out a better way to search for and identify an AWS OIDC Provider that is "owned" by the cluster (the aws sdk is very lacking in this regard. no tag support, lookups only by ARN. [eksctl seems to have this down](https://github.com/weaveworks/eksctl/tree/e44a534ae39ebfaef494ee15cb265107fcc7a6d3/pkg/iam/oidc) so I'll probably mimic what it does). The List functions in `pkg/resources/aws` only take a Cloud and cluster name, so that doesn't seem to be enough to go off of for finding an OIDC provider.
* Create keys.json alongside discovery.json
* Improve api validation 
* Templating the new webhook manifest to include the CA among other things
* docs, testing, etc.

The docs for the pod identity webhook repo have the admission webhook deployment create a CSR on startup and then the deploy instructions manually [kubectl certificate approve](https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/Makefile#L83) the CSR. Since bootstrap manifest builder can't support that, I opted for creating a CA, cert, and key through the normal Kops process, having nodeup pull those down onto the masters and hostPath mount them into the webhook pods. This also means the webhook pods will have to run on the masters which isnt ideal. Any ideas on how kops could approve the CSR automatically?

**Comment by johngmyers** [comment]
I think you'll need a field in the spec for the thumbprint. The admin will have to supply it for an externally hosted issuer. (And I don't like the fact that if Amazon changes the root CA they use for S3 then all our IRSA IDPs will break.)

Kops could approve the CSR by putting an approver in kops-controller. I maintain such a thing: proofpoint/kapprover.

We simply run IRSA without a webhook. It's not required.

**Comment by fred-vogt** [comment]
@rifelpet, @johngmyers - good to see this is being looked into.  This is a huge win for AWS installations.

@johngmyers - I'm testing [amazon-eks-pod-identity-webhook](https://github.com/aws/amazon-eks-pod-identity-webhook) with kops 1.18a3 and k8s 1.18.2 and noticed that the combination of:
* https://github.com/proofpoint/certificate-init-container
* https://github.com/proofpoint/kapprover
No longer works with k8s 1.18
```
level=error ... user not permitted to approve requests with signerName "kubernetes.io/legacy-unknown"
```
* https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md#signers
* https://github.com/kubernetes/enhancements/pull/1400
* https://github.com/kubernetes/kubernetes/pull/88246

**Comment by johngmyers** [comment]
@fred-vogt that should be filed as an issue on the appropriate repo.

**Comment by fred-vogt** [comment]
@johngmyers - good point.  I'll create an issue with your kapprover repo.

For now I want to make the point for others testing with items mentioned here.

**Comment by fred-vogt** [comment]
FWIW: I've got this **self configuring on kops cluster start** mostly from comments in this issue and [Alex Ferener
](https://medium.com/@alex.ferener/iam-roles-for-service-accounts-irsa-on-kops-clusters-a561def8fd2).

I setup both the `mutating admission webhook` and a site-wide `openidc provider` in 
my [multi-account AWS test environment](https://github.com/cicdenv/cicdenv) using:
* manually created OAuth JSON Web Key Set, discovery file, held in AWS Secrets Manager
  * installed via CLI/user-data (pod-identity-webhook running on masters)
* terraform - can correctly compute the issuer cert thumbprint since late July
  * [TLS provider v2.2.0](https://github.com/hashicorp/terraform-provider-tls/releases/tag/v2.2.0) - [tls_certificate](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/data-sources/tls_certificate)
* https://github.com/aws/amazon-eks-pod-identity-webhook
* https://github.com/proofpoint/kapprover
* https://github.com/proofpoint/certificate-init-container

Some thoughts:
```
1 OpenIDConnect Provider for the whole site (Works, but is Bad):
  In practice this means any IAM role in ANY org account
  can be used by ANY other KOPS cluster in ANY other org account.

This isn't quite what I want - I'll be switching to 
1 oidc idp per org account, shared by all KOPS clusters
in the same account.
This is ideal for `my use case` where there should NOT
be a 1-to-1 binding between IAM role and a single KOPS cluster.

TL;DR - for my use case I'd prefer to terraform the 
per account oidc idp and would like KOPS to support
a means of secure delivery.
```

For those looking to test their oidc idp without the webhook:
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: <sa-name>
  namespace: ...
  annotations:
    <annotation-prefix>/role-arn: "<iam-role-arn>"
---
apiVersion: v1
kind: Pod
metadata:
  name: ...
  namespace: ...
spec:
  serviceAccountName: <sa-name>
  containers:
  - name: ...
    image: ...
### Everything below - add manually when NOT using the webhook ###
    env:
    - name: AWS_DEFAULT_REGION
      value: us-west-2
    - name: AWS_REGION
      value: us-west-2
    - name: AWS_ROLE_ARN
      value: "<iam-role-arn>"
    - name: AWS_WEB_IDENTITY_TOKEN_FILE
      value: "<token-mount-path>/token"
    volumeMounts:
    - mountPath: "<token-mount-path>/"
      name: aws-token
  volumes:
  - name: aws-token
    projected:
      sources:
      - serviceAccountToken:
          audience: "sts.amazonaws.com"
          expirationSeconds: 86400
          path: tokenapiVersion: v1
```

**Comment by rrichardson** [comment]
I'm trying to kick the tires on this using the the `ServiceAccountIssuerDIscovery` feature.  I have a more basic architectural question with regards to sync'ing with AWS STS.   If the API Server is providing the discovery/jwks,  I need to provide access to it from STS. 
I think the safest solution here is to set `kubeApiServer:serviceAccountIssuer` to an ELB that is balancing across the masters/apis.  
This doesn't exist yet, so I'm thinking maybe just use a custom DNS CName/Alias that will eventually point to the ELB.  

Am I overthinking this?  
Is there some hook in the Kops pipeline I could use to set this value post-hoc? (that seems unlikely given that it's an API command line param)

**Comment by rifelpet** [comment]
/remove-lifecycle stale

this is in progress

**Comment by Reinkaos** [comment]
This feature would be great for my team too :D.

**Comment by olemarkus** [comment]
Have a look at e.g #11194 and #11016 to see if this covers your need.

**Comment by olemarkus** [comment]
Should anyone have missed the rules, this feature was released with 1.21.
